tephra-core/src/main/java/co/cask/tephra/TransactionContext.java

```diff
@@ -51,11 +54,26 @@ public TransactionContext(TransactionSystemClient txClient, Iterable<Transaction
    * Adds a new transaction-aware to participate in the transaction.
    * @param txAware the new transaction-aware
    */
-  public void addTransactionAware(TransactionAware txAware) {
-    this.txAwares.add(txAware);
-    if (currentTx != null) {
+  public boolean addTransactionAware(TransactionAware txAware) {
+    // If the txAware is newly added, call startTx as well if there is an active transaction
+    boolean added = txAwares.add(txAware);
+    if (added && currentTx != null) {
       txAware.startTx(currentTx);
     }
+    return added;
+  }
+
+  /**
+   * Removes a {@link TransactionAware} to withdraw from participation in the transaction. Withdrawal is only allowed
```

Comment:
Removes a {@link TransactionAware} and withdraws from participation in the transaction. [Not certain if this is what you meant.]